pkg/nsx/services/vpc/vpc.go

```diff
@@ -471,3 +514,230 @@ func (s *VPCService) Cleanup() error {
+			err = errors.New("ipblock not found")
+			log.Error(err, "failed to get public ipblock", "path", cidr)
+			return
+		} else {
+			result = append(result, *ipblock.Cidr)
+		}
+	}
+	return
+}
+
+func (service *VPCService) CreateOrUpdateAVISecurityPolicy(vpc *model.Vpc, namespace string) error {
+	if !service.supportAviAllowRule {
+		return nil
+	}
+	path := strings.Split(*vpc.Path, "/")
+	orgId := path[2]
+	projectId := path[4]
+	ruleId := "allow-all-avi-se"
+	groupId := "all-avi-se-vms"
+	spId := "default-layer3-section"
+
+	if !service.existAVISecurityPolicy(orgId, projectId, *vpc.Id, spId) {
+		return errors.New("avi security policy not found")
+	}
+	allowrule, err := service.getAVIAllowRule(orgId, projectId, *vpc.Id, spId, ruleId)
+	var externalCIDR []string
+	if err == nil {
+		externalCIDR, err = service.getIpblockCidr(vpc.ExternalIpv4Blocks)
+		if err != nil {
+			return err
+		}
+	}
+
+	if allowrule != nil {
+		if !service.needUpdateRule(allowrule, externalCIDR) {
+			log.Info("avi rule no changed, skip update aiv rule")
+			return nil
+		} else {
+			log.Info("avi rule changed", "previous", allowrule.DestinationGroups, "current", externalCIDR)
+		}
+	}
+
+	group, err := service.getAVIGroup(orgId, projectId, *vpc.Id, groupId)
+	if err != nil {
+		log.Info("create avi group", "group", group)
+		group, err = service.createAVIGroup(orgId, projectId, *vpc.Id, groupId)
+		if err != nil {
+			log.Error(err, "failed to create avi group", "group", groupId)
+			return err
+		}
+	}
+	newrule, err := service.buildAVIAllowRule(vpc, externalCIDR, *group.Path, ruleId, projectId)
+	log.Info("creating avi rule", "rule", newrule)
+	if err != nil {
+		log.Error(err, "build avi rule", "rule", newrule)
+		return err
+	}
+
+	err = service.NSXClient.VPCRuleClient.Patch(orgId, projectId, *vpc.Id, spId, *newrule.Id, *newrule)
+	if err != nil {
+		log.Error(err, "create avi rule", "rule", newrule)
+		return err
+	}
+	nsxrule, err := service.NSXClient.VPCRuleClient.Get(orgId, projectId, *vpc.Id, spId, *newrule.Id)
+	if err != nil {
+		log.Error(err, "get avi rule", "rule", nsxrule)
+		return err
+	}
+	service.RuleStore.Add(nsxrule)
+	log.Info("created avi rule")
+	return nil
+}
+
+func (service *VPCService) createAVIGroup(orgId string, projectId string, vpcId string, groupId string) (*model.Group, error) {
+	var group model.Group
+	tags := []model.Tag{
+		{
+			Scope: pointy.String(common.TagScopeCluster),
+			Tag:   pointy.String(service.NSXConfig.Cluster),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVersion),
+			Tag:   pointy.String(strings.Join(common.TagValueVersion, ".")),
+		},
+		{
+			Scope: pointy.String(common.TagScopeVPCCRUID),
+			Tag:   pointy.String(vpcId),
+		},
+		{
+			Scope: pointy.String(common.TagScopeGroupType),
+			Tag:   pointy.String("avi"),
+		},
+	}
+	group.Tags = tags
+	resource_type := "Condition"
+	scope_op := "EQUALS"
+	operator := "EQUALS"
+	member_type := "VpcSubnet"
+	value := "AVI_SUBNET_LB|"
+	key := "Tag"
```

Comment:
#L649 to #L655, no need to define these variables, just pass as params,

Reply:
done